velox/expression/CastExpr.cpp

```diff
@@ -39,40 +39,34 @@ namespace {
 /// @tparam From The expression type
 /// @param row The index of the current row
 /// @param input The input vector (of type From)
-/// @param resultFlatVector The output vector (of type To)
+/// @param result The output vector (of type To)
 /// @return False if the result is null
 template <typename To, typename From, bool Truncate>
 void applyCastKernel(
     vector_size_t row,
-    const BaseVector& input,
-    FlatVector<To>* resultFlatVector,
+    const SimpleVector<From>* input,
+    FlatVector<To>* result,
     bool& nullOutput) {
-  auto* inputVector = input.asUnchecked<SimpleVector<From>>();
-
   // Special handling for string target type
   if constexpr (CppToType<To>::typeKind == TypeKind::VARCHAR) {
-    if (nullOutput) {
-      resultFlatVector->setNull(row, true);
-    } else {
-      auto output =
-          util::Converter<CppToType<To>::typeKind, void, Truncate>::cast(
-              inputVector->valueAt(row), nullOutput);
+    auto output =
+        util::Converter<CppToType<To>::typeKind, void, Truncate>::cast(
+            input->valueAt(row), nullOutput);
+    if (!nullOutput) {
       // Write the result output to the output vector
-      auto writer = exec::StringWriter<>(resultFlatVector, row);
```

Comment:
This logic should be conditioned on 'nullOutput' being false. It just happens so that Converter::cast never sets 'nullOutput' to true.